### ISSUE TITLE
Unrecognized KeyValue (TryReadAsBinary fails)

### ISSUE THREAD

**Issue by JustArchi** [SteamRE/SteamKit#376]
Since a few hours a function that worked just fine for 2 years, responsible for reading `KeyValue` returned by Steam suddenly stopped working and I'm trying to find out why.

During investigation I found out that that SK2 is unable to read it, while normal Steam client does, so obviously Valve changed something during last maintenance that happened a few hours ago and new KV is no longer possible to read, so I decided to file a bug report. It's possible that this change affects more than a single thing, but I was able to reproduce it with `purchase_receipt_info` of `ClientPurchaseResponse (763)`. It could be temporary change, but I doubt it since nothing like that happened in last 2 years.

During debugging I found out that SK2 properly reads binary to some point, after which probably there is some invalid structure and reading it further is impossible.

![image](https://cloud.githubusercontent.com/assets/1069029/24179149/a67d26b6-0ead-11e7-8cf2-2d89ce5670c0.png)

^ Only 4 items read, should be much more.

Please feel free to check with following KV (saved in hexadecimal form, as recorded by Nethook2):

```
004D6573736167654F626A65637400075472616E73616374696F6E494400FFFFFFFFFFFFFFFF027061796D656E744D6574686F640001000000025075726368617365537461747573000200000002526573756C7444657461696C00090000000A4261736550726963650000000000000000000A546F74616C446973636F756E740000000000000000000A5461780000000000000000000A5368697070696E670000000000000000000243757272656E6379436F64650000000000025472616E73616374696F6E54696D650091CED158025061636B616765494400FFFFFFFF024C696E654974656D436F756E740001000000006C696E656974656D7300003000075472616E73616374696F6E494400FFFFFFFFFFFFFFFF025061636B61676549440016730100014974656D4465736372697074696F6E004D616347756666696E000808014572726F72486561646C696E650000014572726F72537472696E670000014572726F724C696E6B546578740000014572726F724C696E6B55524C00000808
```

In case you need extra code, you can find it **[here](https://github.com/JustArchi/ArchiSteamFarm/blob/5e5e4d11f8db01f5b3ac102e6d4cef1fa88feeb5/ArchiSteamFarm/ArchiHandler.cs#L396)**.

Thank you a lot for looking into it!

**Comment by Netshroud** [comment]
I need to do a further analysis but it looks like Valve may have introduced a new binary keyvalue type - `0x0A` (10) describing a signed 64-bit integer.

**Comment by Netshroud** [comment]
For everyone else reading this there also appears to be `9` which is another int, but I can't tell yet how it differs to `5`.

**Comment by JustArchi** [comment]
Thanks a lot for solving it so quickly @Netshroud ! Do you plan to release a small maintenance release (1.8.2?) with this fix in? It's crucial for me since old code stopped working - I can always compile myself, but I think that it will be wanted by far more people.

In any case I'm very grateful that you managed to fix it ❤️

**Comment by psychonic** [comment]
9 is TYPE_BINARY_STRING. Not sure how it's used.

**Comment by Netshroud** [comment]
@JustArchi I was gonna leave it for 24 hours or so while assessing type 9.

@psychonic: Any idea if it's different on the wire to Pointer/Colour/Int32?

Interestingly the latest steamclient binary I have now treats Pointer now as UInt64, not at (U)Int32... 😕

**Comment by Netshroud** [comment]
@JustArchi I've just published 1.8.2 with the fix for this.

I'm still concerned about `TYPE_POINTER` and `TYPE_BINARY_STRING`, but we can worry about that when Valve actually use them (publicly).

`TYPE_POINTER` seems to be `uint64` in latest build of Steam, not `int32`.

`TYPE_BINARY_STRING` appears to be length-prefixed data - a `int32` (signed) followed by that number of bytes.

I'm hesitant to change/add the code for either of those without a valid sample.
